ext/mbstring/mbstring.c

```diff
@@ -3486,16 +3462,11 @@ static zend_string* html_numeric_entity_decode(zend_string *input, const mbfl_en
 					}
 				}
 			}
-		} else if (p2 == wchar_buf + out_len) {
+		} else if ((p2 == wchar_buf + out_len) && in_len) {
 			/* Corner case: & at end of buffer */
-			if (in_len) {
-				wchar_buf[0] = '&';
-				wchar_buf_offset = 1;
-				goto process_converted_wchars;
-			} else {
-				*converted++ = '&';
```

Comment:
Correct handling of this case falls out automatically from the code below; no need to handle it with a dedicated `if` clause.